### PR TITLE
Diagram: added diagram writing for generator with diagram model

### PR DESCRIPTION
**Pull request by PerninGuillaume**
Signed-off-by: Guillaume Pernin <guillaume.pernin@rte-france.com>

### CODE REVIEW

File: sources/Outputs/include/Diagram.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ public:
+  /**
+  * @brief Dyd definition to provide informations to build the Dyd file
+  */
+  struct DiagramDefinition {
+    /**
+     * @brief Constructor
+     *
+     * @param base the basename for current file (corresponds to filepath basename)
+     * @param filepath the filepath of the dyd file to write
```
Comment (rosiereflo) [inline sources/Outputs/include/Diagram.h:40]: please update the comment

File: sources/Outputs/include/Diagram.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    std::string basename;                               ///< basename for file
+    std::string filename;                               ///< filepath for file to write
+    std::vector<algo::GeneratorDefinition> generators;  ///< generators found
+  };
+
+  /**
+   * @brief Constructor
+   *
+   * @param def the dyd definition
```
Comment (rosiereflo) [inline sources/Outputs/include/Diagram.h:54]: Same as above

File: sources/Outputs/include/Diagram.h [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+   */
+  explicit Diagram(DiagramDefinition&& def);
+
+  /**
+   * @brief Write the dyd file
```
Comment (rosiereflo) [inline sources/Outputs/include/Diagram.h:59]: same as above (copy paste is bad! ;))

File: sources/Outputs/include/Diagram.h [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+   */
+  void write();
+
+ private:
+  void write_table(const algo::GeneratorDefinition& generator, std::stringstream& buffer, bool is_table_qmin);
+  DiagramDefinition def_;  ///< Dyd file information
```
Comment (rosiereflo) [inline sources/Outputs/include/Diagram.h:65]: same

File: sources/Outputs/src/Diagram.cpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    std::sort(generator.points.begin(), generator.points.end(),
+              [](const DYN::GeneratorInterface::ReactiveCurvePoint& gen1, const DYN::GeneratorInterface::ReactiveCurvePoint& gen2) -> bool {
+                return gen1.p < gen2.p;
+              });
+
+    write_table(generator, buffer, true);
+    write_table(generator, buffer, false);
+  }
+
+  if (not empty) {
```
Comment (lecourtoisflo) [inline sources/Outputs/src/Diagram.cpp:52]: Do not respect sonar rule https://rules.sonarsource.com/cpp/RSPEC-3659. same for the 'and' above

File: sources/Outputs/src/Diagram.cpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  else
+    buffer << constants::diagramMaxTableSuffix << '(';
+  std::size_t number_lines = generator.points.empty() ? 2 : generator.points.size();
+  //The number of lines is 2 when there are no points
+  buffer << number_lines;
+  buffer << ",2)";
+
+  if (generator.points.empty()) {
+    double q_value = is_table_qmin ? generator.qmin : generator.qmax;
+    buffer << '\n' << generator.pmin / 100 << " " << q_value / 100 << '\n';
```
Comment (lecourtoisflo) [inline sources/Outputs/src/Diagram.cpp:76]: put the 100 in a class constant as it is used many times

File: sources/Outputs/src/Diagram.cpp [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        generator.model != algo::GeneratorDefinition::ModelType::WITH_IMPEDANCE_DIAGRAM_PQ_SIGNALN)
+      continue;
+    empty = false;
+
+    std::sort(generator.points.begin(), generator.points.end(),
+              [](const DYN::GeneratorInterface::ReactiveCurvePoint& gen1, const DYN::GeneratorInterface::ReactiveCurvePoint& gen2) -> bool {
+                return gen1.p < gen2.p;
+              });
+
+    write_table(generator, buffer, true);
```
Comment (rosiereflo) [inline sources/Outputs/src/Diagram.cpp:48]: Please respect the coding style (writeTable)

Comment (rosiereflo) [reply sources/Outputs/src/Diagram.cpp:48]: Also, please declare const qmin = true, const qmax = false and use them in the calls to improve lisibility

File: sources/Outputs/src/Diagram.cpp (hunk 5 above)
Comment (rosiereflo) [inline sources/Outputs/src/Diagram.cpp:52]: !empty

**Review by lecourtoisflo — CHANGES_REQUESTED** [review]

File: sources/Outputs/src/Diagram.cpp [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+  if (not empty) {
+    //Put buffer into file
+    std::ofstream ofs(def_.filename, std::ofstream::out);
+    ofs << buffer.str();
+    ofs.close();
+  }
+}
+
+void
+Diagram::write_table(const algo::GeneratorDefinition& generator, std::stringstream& buffer, bool is_table_qmin) {
```
Comment (rosiereflo) [inline sources/Outputs/src/Diagram.cpp:61]: coding style: isTableQMin

File: sources/Outputs/src/Diagram.cpp [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+  buffer << "\ndouble ";
+  auto hash = constants::hash(generator.id);
+  buffer << hash;
+  if (is_table_qmin)
+    buffer << constants::diagramMinTableSuffix << '(';
+  else
+    buffer << constants::diagramMaxTableSuffix << '(';
+  std::size_t number_lines = generator.points.empty() ? 2 : generator.points.size();
+  //The number of lines is 2 when there are no points
```
Comment (rosiereflo) [inline sources/Outputs/src/Diagram.cpp:70]: typo: there is no point

File: sources/Outputs/src/Diagram.cpp [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+  buffer << number_lines;
+  buffer << ",2)";
+
+  if (generator.points.empty()) {
+    double q_value = is_table_qmin ? generator.qmin : generator.qmax;
+    buffer << '\n' << generator.pmin / 100 << " " << q_value / 100 << '\n';
+    buffer << generator.pmax / 100 << " " << q_value / 100;
+  } else {
+    for (const DYN::GeneratorInterface::ReactiveCurvePoint& point : generator.points) {
```
Comment (rosiereflo) [inline sources/Outputs/src/Diagram.cpp:79]: auto?

File: sources/Outputs/src/Diagram.cpp [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+      double q_value = is_table_qmin ? point.qmin : point.qmax;
+      buffer << '\n' << point.p / 100 << " " << q_value / 100;
+    }
+  }
+}
+
+}  // namespace outputs
+}  // namespace dfl
```
Comment (rosiereflo) [inline sources/Outputs/src/Diagram.cpp:87]: Missing newline

File: tests/outputs/CMakeLists.txt [hunk 12]
```diff
@@ -15,3 +15,6 @@ target_link_libraries(TestDyd DynaFlowLauncher::outputs)
 
 DEFINE_TEST(TestPar OUTPUTS)
 target_link_libraries(TestPar DynaFlowLauncher::outputs)
+
+DEFINE_TEST(TestDiagram OUTPUTS)
+target_link_libraries(TestDiagram DynaFlowLauncher::outputs)
```
Comment (rosiereflo) [inline tests/outputs/CMakeLists.txt:20]: Missing new line

File: tests/outputs/reference/TestPar/TestPar.par [hunk 13]
```diff
@@ -1,14 +1,18 @@
 <?xml version="1.0" encoding="ISO-8859-1" standalone="no"?>
 <parametersSet xmlns="http://www.rte-france.com/dynawo">
-  <set id="G2">
+  <set id="10160083025018531003">
     <par name="generator_KGover" type="DOUBLE" value="1"/>
-    <par name="generator_QMax0" type="DOUBLE" value="31"/>
+    <par name="generator_QMax0" type="DOUBLE" value="41"/>
```
Comment (rosiereflo) [inline tests/outputs/reference/TestPar/TestPar.par:5]: Why did this change?

File: tests/outputs/TestDiagram.cpp [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+
+#include "Diagram.h"
+#include "Tests.h"
+
+#include <boost/filesystem.hpp>
+
+TEST(Diagram, writeWithCurvePoint) {
+  using dfl::algo::GeneratorDefinition;
+
+  std::string basename = "TestDiagram";
```
Comment (rosiereflo) [inline tests/outputs/TestDiagram.cpp:19]: Please also test the empty case

**Review by rosiereflo — CHANGES_REQUESTED** [review]
Looks globally good!

File: sources/Outputs/src/Diagram.cpp (hunk 7 above)
Comment (lecourtoisflo) [reply sources/Outputs/src/Diagram.cpp:48]: or an enum instead of a boolean

File: tests/outputs/CMakeLists.txt (hunk 12 above)
Comment (lecourtoisflo) [reply tests/outputs/CMakeLists.txt:20]: Not a requirement for cmakelists, it is the case for other cmakelists

**Review by lecourtoisflo — CHANGES_REQUESTED** [review]

File: sources/Outputs/src/Diagram.cpp (hunk 5 above)
Comment (PerninGuillaume) [reply sources/Outputs/src/Diagram.cpp:52]: I don't really understand this rule, what is more readable && or "and", in my opinion it is obviously the latter one. Why would the C++ have introduced alternative operators if no one can use it ?

Comment (rosiereflo) [reply sources/Outputs/src/Diagram.cpp:52]: 1 argument: we want to be homogeneous on our coding style in the project, so I agree with Florent we should use the old ones (this is project specific, another project might accept the new operators)

File: sources/Outputs/src/Diagram.cpp (hunk 7 above)
Comment (PerninGuillaume) [reply sources/Outputs/src/Diagram.cpp:48]: Shoud I also change the variables names like is_table_qmin to follow the naming convention ?

Comment (PerninGuillaume) [reply sources/Outputs/src/Diagram.cpp:48]: Oh didn't read far enough, I had not seen your comment farther down

File: sources/Outputs/src/Diagram.cpp (hunk 9 above)
Comment (PerninGuillaume) [reply sources/Outputs/src/Diagram.cpp:70]: From what I found, I think both can be written but the first is more preferable 
[https://english.stackexchange.com/questions/1931/there-are-no-comments-vs-there-is-no-comment](url)

File: sources/Outputs/src/Diagram.cpp (hunk 10 above)
Comment (PerninGuillaume) [reply sources/Outputs/src/Diagram.cpp:79]: I don't like it, it may be longer to write the full type, however any doubt you could have about the type when there is a auto are gone with the full type.

File: tests/outputs/reference/TestPar/TestPar.par (hunk 13 above)
Comment (PerninGuillaume) [reply tests/outputs/reference/TestPar/TestPar.par:5]: In fact the two generators have been swapped in the ordering. I don't have the hand on this but apparently the xmlWriter reorder the generators based on their id. Before G2 was before G3 in the xml, with the hash, G2 became 6320417593396202500 and G3 became 10160083025018531003. Therefore G3 took the place of G2 and the generator_QMax0 values have been swapped

File: sources/Outputs/include/Diagram.h [hunk 15]
```diff
@@ -51,18 +52,20 @@ class Diagram {
...
    */
   void write();
 
  private:
-  void write_table(const algo::GeneratorDefinition& generator, std::stringstream& buffer, bool is_table_qmin);
-  DiagramDefinition def_;  ///< Dyd file information
+  enum class Tables { TABLE_QMIN = 0, TABLE_QMAX };
+  void writeTable(const algo::GeneratorDefinition& generator, std::stringstream& buffer, Tables table);
+  DiagramDefinition def_;  ///< Diagram file information
+  std::size_t divisorFactor_ = 100;
```
Comment (rosiereflo) [inline sources/Outputs/include/Diagram.h:68]: Please use a const local number instead of having a const attribute

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: sources/Outputs/include/Diagram.h [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+  /**
+   * @brief Write the Diagram file
+   */
+  void write();
+
+ private:
+  enum class Tables { TABLE_QMIN = 0, TABLE_QMAX };
+  void writeTable(const algo::GeneratorDefinition& generator, std::stringstream& buffer, Tables table);
+  DiagramDefinition def_;  ///< Diagram file information
+  std::size_t divisorFactor_ = 100;
```
Comment (lecourtoisflo) [inline sources/Outputs/include/Diagram.h:68]: divisor factory should be static constexpr. std:: o std::size_t is not required

File: sources/Outputs/include/Diagram.h [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+   */
+  explicit Diagram(DiagramDefinition&& def);
+
+  /**
+   * @brief Write the Diagram file
+   */
+  void write();
+
+ private:
+  enum class Tables { TABLE_QMIN = 0, TABLE_QMAX };
```
Comment (lecourtoisflo) [inline sources/Outputs/include/Diagram.h:65]: Add doxygen comments

File: sources/Outputs/src/Constants.h [hunk 18]
```diff
@@ -31,6 +31,9 @@ const std::string diagramMaxTableSuffix{"_tableqmax"};              ///< Suffix
 const std::string diagramMinTableSuffix{"_tableqmin"};              ///< Suffix for the table name for qmin in diagram file
 const std::string signalNGeneratorParId{"signalNGenerator"};        ///< PAR id for generators using signal N
 const std::string impSignalNGeneratorParId{"impSignalNGenerator"};  ///< PAR id for generators using signal N with impendance
+static inline std::size_t hash(std::string str) {
```
Comment (lecourtoisflo) [inline sources/Outputs/src/Constants.h:34]: Add doxygen comments

**Review by lecourtoisflo — CHANGES_REQUESTED** [review]

File: sources/Outputs/src/Diagram.cpp [hunk 19]
```diff
@@ -70,14 +70,15 @@ Diagram::writeTable(const algo::GeneratorDefinition& generator, std::stringstrea
   buffer << numberLines;
   buffer << ",2)";
 
+  int divisorFactor = 100;
```
Comment (rosiereflo) [inline sources/Outputs/src/Diagram.cpp:73]: Please add a const identifier and a static

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: sources/Outputs/src/Diagram.cpp (hunk 19 above)
Comment (lecourtoisflo) [reply sources/Outputs/src/Diagram.cpp:73]: > Please add a const identifier and a static

const, yes, static not necessary here the function is called ony once

Comment (rosiereflo) [reply sources/Outputs/src/Diagram.cpp:73]: OK then, const will be enough :)

**Review by rosiereflo — APPROVED** [review]

File: sources/Outputs/src/Constants.h [hunk 20]
```diff
@@ -31,6 +31,12 @@ const std::string diagramMaxTableSuffix{"_tableqmax"};              ///< Suffix
 const std::string diagramMinTableSuffix{"_tableqmin"};              ///< Suffix for the table name for qmin in diagram file
 const std::string signalNGeneratorParId{"signalNGenerator"};        ///< PAR id for generators using signal N
 const std::string impSignalNGeneratorParId{"impSignalNGenerator"};  ///< PAR id for generators using signal N with impendance
+  /**
+   * @brief Return a hash number from a string as input
+   */
+static inline std::size_t hash(std::string str) {
```
Comment (lecourtoisflo) [inline sources/Outputs/src/Constants.h:37]: input string should be const &

Comment (lecourtoisflo) [inline sources/Outputs/src/Constants.h:37]: functions are variables should be in different sections

File: sources/Outputs/src/Par.cpp [hunk 21]
```diff
@@ -128,12 +127,14 @@ Par::writeConstantSets(unsigned int nb_generators) {
...
-  auto set = parameters::ParametersSetFactory::newInstance(def.id);
-
   if (def.model == algo::GeneratorDefinition::ModelType::SIGNALN || def.model == algo::GeneratorDefinition::ModelType::WITH_IMPEDANCE_SIGNALN) {
     // already processed by constant
     return nullptr;
   }
+  std::size_t hashId = constants::hash(def.id);
+  std::string hashIdStr = std::to_string(hashId);
+  auto set = parameters::ParametersSetFactory::newInstance(hashIdStr);
+  //change def.id to the hash
```
Comment (lecourtoisflo) [inline sources/Outputs/src/Par.cpp:137]: This coment is irrelevant for final code. change it something like "use the hash id in exported files to prevent use of non-ascii characters"

File: sources/Outputs/include/Diagram.h [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+ * @brief Dynaflow launcher Diagram file writer header file
+ *
+ */
+
+#pragma once
+
+#include "Algo.h"
+
+#include <string>
+#include <unordered_map>
```
Comment (lecourtoisflo) [inline sources/Outputs/include/Diagram.h:23]: unused header

File: sources/Outputs/include/Diagram.h [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+  struct DiagramDefinition {
+    /**
+     * @brief Constructor
+     *
+     * @param base the basename for current file (corresponds to filepath basename)
+     * @param filepath the filepath of the diagram file to write
+     * @param gens generators definition coming from algorithms
+     */
+    DiagramDefinition(const std::string& base, const std::string& filepath, const std::vector<algo::GeneratorDefinition>& gens) :
+        basename{base}, filename{filepath}, generators{gens} {}
```
Comment (lecourtoisflo) [inline sources/Outputs/include/Diagram.h:45]: coding style not espected. did you use the formatter ?

**Review by lecourtoisflo — CHANGES_REQUESTED** [review]

File: tests/outputs/TestDiagram.cpp [hunk 24]
```diff
@@ -0,0 +1,139 @@
...
+
+  DiagramWriter.write();
+  std::string filePath = outputDir + "/" + filename;
+  try {
+    std::ifstream stream(filePath);
+    ASSERT_TRUE(stream.fail()) << "The Diagram file has been written to even though there are no generators which model requires it." << std::endl;
+  } catch (std::exception& e) {
+    ASSERT_TRUE(false) << "An IO error has occured" << std::endl;
+  }
+}
```
Comment (gautierbureau) [inline tests/outputs/TestDiagram.cpp:139]: newline

File: sources/Outputs/src/Diagram.cpp [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+namespace outputs {
+
+Diagram::Diagram(DiagramDefinition&& def) : def_{std::forward<DiagramDefinition>(def)} {}
+
+void
+Diagram::write() {
+  //We don't want to create the file if there are no generators diagrams to write
+  //maybe use stdbuf for using << and not .append(..()))
+  std::stringstream buffer;
+  buffer << "#1";
```
Comment (gautierbureau) [inline sources/Outputs/src/Diagram.cpp:34]: add a comment to precise the meaning of 1 here

**Review by gautierbureau — CHANGES_REQUESTED** [review]

File: sources/Outputs/src/Diagram.cpp (hunk 25 above)
Comment (PerninGuillaume) [reply sources/Outputs/src/Diagram.cpp:34]: You mean that modelica absolutely requires it ?

File: tests/outputs/TestDiagram.cpp [hunk 26]
```diff
@@ -134,6 +134,6 @@ TEST(Diagram, writeEmpty) {
     std::ifstream stream(filePath);
     ASSERT_TRUE(stream.fail()) << "The Diagram file has been written to even though there are no generators which model requires it." << std::endl;
   } catch (std::exception& e) {
-    ASSERT_TRUE(false) << "An IO error has occured" << std::endl;
+    ASSERT_TRUE(false) << "An IO error has occured\n" << std::endl;
```
Comment (lecourtoisflo) [inline tests/outputs/TestDiagram.cpp:137]: why add another end of line character ?

**Review by lecourtoisflo — CHANGES_REQUESTED** [review]

File: tests/outputs/TestDiagram.cpp (hunk 26 above)
Comment (PerninGuillaume) [reply tests/outputs/TestDiagram.cpp:137]: Ah gautier induced me in error, see his remark above

File: sources/Outputs/src/Diagram.cpp (hunk 25 above)
Comment (gautierbureau) [reply sources/Outputs/src/Diagram.cpp:34]: I don't know the format of the file and there is no documentation at the moment on it, so add a comment to give the information to a developer

File: tests/outputs/TestDiagram.cpp (hunk 26 above)
Comment (gautierbureau) [reply tests/outputs/TestDiagram.cpp:137]: newline at the end of the file not here

**Review by lecourtoisflo — APPROVED** [review]

**Comment by gautierbureau** [comment]
already review it and for me there is still one unresolved subject

**Comment by rosiereflo** [comment]
> already reviewed it and for me there is still one unresolved subject
What is the remaining subject? both opened discussions were addressed as far as I understand

**Comment by gautierbureau** [comment]
> > already reviewed it and for me there is still one unresolved subject
> 
> What is the remaining subject? both opened discussions were addressed as far as I understand

It is ok now I did not see he added the comment because it was below my conversation and I had to see it in files changed and not on this page, my bad

**Review by gautierbureau — APPROVED** [review]

**Comment by rosiereflo** [comment]
thx!
